pkg/apis/storage/types.go

```diff
@@ -215,3 +215,174 @@ const (
 	// binding will occur during Pod scheduing.
 	VolumeBindingWaitForFirstConsumer VolumeBindingMode = "WaitForFirstConsumer"
 )
+
+// VolumePluginMechanism is the type of mechanism components should use for volume operations
+type VolumePluginMechanism string
+
+const (
+	// VolumePluginMechanismInTree means components should use the in-tree plugin for volume operations
+	VolumePluginMechanismInTree VolumePluginMechanism = "in-tree"
+	// VolumePluginMechanismCSI means components should use the CSI Driver for volume operations
+	VolumePluginMechanismCSI VolumePluginMechanism = "csi"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIDriver captures information about a Container Storage Interface (CSI)
+// volume driver deployed on the cluster.
+// CSIDriver objects are non-namespaced.
+type CSIDriver struct {
+	metav1.TypeMeta
+
+	// Standard object metadata.
+	// metadata.Name indicates the name of the CSI driver that this object
+	// refers to; it MUST be the same name returned by the CSI GetPluginName()
+	// call for that driver.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ObjectMeta
+
+	// Specification of the CSI Driver.
+	Spec CSIDriverSpec
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIDriverList is a collection of CSIDriver objects.
+type CSIDriverList struct {
+	metav1.TypeMeta
+
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// items is the list of CSIDriver
+	Items []CSIDriver
+}
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+	// attachRequired indicates this CSI volume driver requires an attach
+	// operation (because it implements the CSI ControllerPublishVolume()
+	// method), and that Kubernetes should call attach and wait for any attach
+	// operation to complete before proceeding to mounting.
+	// If value is not specified, default is false -- meaning attach will not be
+	// called.
+	// +optional
+	AttachRequired *bool
+
+	// If specified, podInfoRequiredOnMount indicates this CSI volume driver
+	// requires additional pod information (like podName, podUID, etc.) during
+	// mount operations.
+	// If value is not specified, pod information will not be passed on mount.
+	// If value is set to a valid version, Kubelet will pass pod information as
+	// VolumeAttributes in the CSI NodePublishVolume() calls.
+	// Supported versions:
+	// Version "v1" will pass the following ValueAttributes
+	// "csi.storage.k8s.io/pod.name": pod.Name
+	// "csi.storage.k8s.io/pod.namespace": pod.Namespace
+	// "csi.storage.k8s.io/pod.uid": string(pod.UID)
+	// +optional
+	PodInfoOnMountVersion *string
```

Comment:
Ditto about using string pointer here. Is empty string same as not setting the value? if yes - why not keep it plain string?

Reply:
IMO - now that we are moving this in-tree, I am not super sure about versioning scheme being specified here makes sense.  What does "v1" even mean?

Reply:
@davidz627 can you answer the question about "v1"?

Reply:
what `v1` are we referring to here? The APIs we're moving (CSINodeInfo, CSIDriverSpec) should all be in `v1beta1` and the meaning of that is defined here:
https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning

Reply:
We used to set `PodInfoOnMountVersion` to `v1` in CRs and that triggers sending of pod information on publish call. I am just wondering if we still want this string's value to be `v1` ? It doesn't seem very intuitive.

Reply:
cc @jsafrane 

This "v1" refers to the set of pod metadata that we choose to give to the csi driver

Reply:
Also should this prefix "csi.storage.k8s.io" be changed to "storage.k8s.io"?

Reply:
The prefix should still remain as "csi.storage.k8s.io".  This prefix is our reserved prefix for CreateVolume parameters that the Kubernetes injects to distinguish it from storage provider parameters.

Reply:
@jsafrane can you confirm that "v1" should stay the same here?

Reply:
> The prefix should still remain as "csi.storage.k8s.io". This prefix is our reserved prefix for CreateVolume parameters that the Kubernetes injects to distinguish it from storage provider parameters.

I thought we were referring to the group name for the API, apologies

Reply:
I am still not sure if this should be called "PodInfoMountVersion".  The `v1` string we are setting is entirely arbitrary and has no meaning in itself. Have we considered renaming this field to something like `AttributePolicy` or something else with enum type, so as k8s can reason about values to expect here.